pkg/controller/nodeimage/nodeimage_controller.go

```diff
@@ -0,0 +1,468 @@
+		} else {
+			klog.Infof("Finish to process NodeImage %v, elapsedTime %v", request.Name, time.Since(start))
+		}
+	}()
+
+	// Fetch the Node
+	node := &v1.Node{}
+	err = r.Get(context.TODO(), request.NamespacedName, node)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			return reconcile.Result{}, fmt.Errorf("failed to get node %s: %v", request.NamespacedName.Name, err)
+		}
+		node = nil
+	}
+
+	// Fetch the NodeImage
+	nodeImage := &appsv1alpha1.NodeImage{}
+	err = r.Get(context.TODO(), request.NamespacedName, nodeImage)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			return reconcile.Result{}, fmt.Errorf("failed to get nodeimage %s: %v", request.NamespacedName.Name, err)
+		}
+		nodeImage = nil
+	}
+
+	// If Node not exists or has been deleted
+	if node == nil || node.DeletionTimestamp != nil {
+		// All been deleted
+		if nodeImage == nil || nodeImage.DeletionTimestamp != nil {
+			return reconcile.Result{}, nil
+		}
+
+		if _, ok := nodeImage.Labels[fakeLabelKey]; !ok {
+			// Delete nodeImage
+			if err = r.Delete(context.TODO(), nodeImage); err != nil {
+				return reconcile.Result{}, fmt.Errorf("failed to delete nodeimage %v, err: %v", nodeImage.Name, err)
+			}
+			return reconcile.Result{}, nil
+		}
+	}
+
+	// If Node exists, we should create a NodeImage
+	if nodeImage == nil {
+		nodeImage = &appsv1alpha1.NodeImage{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: request.Name,
+			},
+			Spec: appsv1alpha1.NodeImageSpec{},
+		}
+		if err = r.Create(context.TODO(), nodeImage); err != nil {
+			return reconcile.Result{}, fmt.Errorf("failed to create nodeimage %v, err: %v", nodeImage.Name, err)
+		}
+		klog.Infof("Successfully create nodeimage %s", nodeImage.Name)
+	}
+
+	if nodeImage.DeletionTimestamp != nil {
+		return reconcile.Result{}, nil
+	}
+
+	duration := &requeueduration.Duration{}
+	if modified, err := r.updateNodeImage(nodeImage.Name, node, duration); err != nil {
+		return reconcile.Result{}, err
+	} else if modified {
+		return reconcile.Result{}, nil
+	}
+	if err = r.updateNodeImageStatus(nodeImage, duration); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	res = reconcile.Result{}
+	res.RequeueAfter, requeueMsg = duration.GetWithMsg()
+	if res.RequeueAfter > 0 && res.RequeueAfter < minRequeueDuration {
+		// 3~5s
+		res.RequeueAfter = minRequeueDuration + time.Duration(rand.Int31n(2000))*time.Millisecond
+	}
+	return res, nil
+}
+
+func (r *ReconcileNodeImage) updateNodeImage(name string, node *v1.Node, duration *requeueduration.Duration) (bool, error) {
+	var modified bool
+	var messages []string
+	tmpDuration := &requeueduration.Duration{}
+	err := retry.RetryOnConflict(retry.DefaultBackoff, func() (err error) {
+		nodeImage := &appsv1alpha1.NodeImage{}
+		err = r.Get(context.TODO(), types.NamespacedName{Name: name}, nodeImage)
+		if err != nil {
+			return err
+		}
+
+		if node != nil {
```

Comment:
extract a function of update nodeimage spec to make body of RetryOnConflict more clear

Reply:
done